comp.go

```diff
@@ -2,47 +2,40 @@ package vstore
 
 import (
 	"github.com/gopherjs/vecty"
-	"github.com/gopherjs/vecty/elem"
 )
 
-// ComponentConstructor is a function that connects a Component's props
-// to the Store's State. This will get called everytime a Store updates,
-// therefore you can just use it instantiate your Component struct and
-// give it the updated property from Store.GetState().
-type ComponentConstructor func(s *Store) vecty.Component
-
-// NewComponent returns a vecty Component that will rerender everytime
-// the store is updated.
-func (s *Store) NewComponent(cc ComponentConstructor) vecty.Component {
-	c := &comp{cc: cc, s: s}
-
-	return c
+// StoreComponent must be satisfied by components to connect to a store.
+// Also, vecty.Component is embedded to preserve component behavior.
+type StoreComponent interface {
+	vecty.Component
+	Connect(store Store)
 }
 
-type comp struct {
-	vecty.Core
-	cc    ComponentConstructor
-	s     *Store
-	unsub func()
-	test  interface{}
+// storeComponent wraps a component to add store pubsub logic.
+// Also, StoreComponent is embedded to preserve component behavior.
+type storeComponent struct {
+	StoreComponent
+	store *store
 }
 
-func (c *comp) Render() *vecty.HTML {
-	return elem.Div(
-		c.cc(c.s),
-	)
+// Connect links the store to a given component and returns the component.
+func (s *store) Connect(comp StoreComponent) StoreComponent {
+	comp.Connect(s)
+	return &storeComponent{StoreComponent: comp, store: s}
 }
 
-func (c *comp) Mount() {
-	c.unsub = c.s.Subscribe(c.callback)
+// Mount subscribes the component to the store and optionally calls mount.
+func (c *storeComponent) Mount() {
+	c.store.sub(c)
+	if mounter, ok := c.StoreComponent.(vecty.Mounter); ok {
```

Comment:
this is neat. However, we would have to support and keep up with all the life-cycle hooks that vecty will provide. Think react's `componentWillMount`, `componentShouldUpdate`, and `componentWillReceiveProps`. It would be an annoyance to have to keep an eye on every new vecty life-cycle feature. 

The simple workaround is this : 

```go
func (c *storeComponent) Render() vecty.HTML {
    return elem.Div(c.StoreComponent)
}
```

this will cause all the life cycle hooks of the child component to trigger. Note that we cannot do `return c.StoreComponent.Render()` so our higher order component has to wrap the child with an element.

Reply:
I haven't had a chance to provide examples there as I've been working on keyed children (now merged), but I filed gopherjs/vecty#148 exactly because of this sort of use-case.

Reply:
@pdf that would be ideal! As of now, the extra `div` doesn't bother me because it means `vstore` doesn't need to keep up with vecty's lifecycles and the user experience stays fluid.

Reply:
Introducing an additional (and unmodifiable!) DOM element can cause significant problems.  I think there's a general design problem here introduced by trying to follow the redux example too closely.

I don't see any reason to wrap components for this functionality, instead a new interface for components to implement that describes a callback for the store to hit would allow components to register themselves to the store, without needing to retain a reference to the store, and without needing to wrap components.

It's super-late here now, so I've got to get some sleep, will have to follow up in the coming days.

Reply:
@pdf you should define and give examples of what "significant problems" mean.

Reply:
Adding extra structural elements breaks layouts that rely on specific nesting or CSS.

Reply:
css is cascading, so it shouldn't be a problem as long as a user can specify what element the wrapping parent is, and can even pass it classes or props. which definitely should be included in vstore as a feature. 

All of that discussion, however, would be solved if that issue you mentioned initially is resolved on the vecty side. How likely is that @pdf ?

Reply:
It's not always so simple, particularly when dealing with third-party CSS frameworks and the like, and can be difficult to deal with when working with grids/flexbox/etc.

Changing the component render return would be a moderately complex change, so I'm not sure if it will happen or not at this stage.  I need to review the current issue priorities before I can investigate in more depth and work out if the cost (time/complexity) is worth it.

Reply:
Keep in mind we are using embedding not composition aka wrapping. This allows the component will continue to satisfy any behavior it wants to. Therefore, we will not have to keep up with all the vecty lifecycle hooks, just the ones we decorate.

Reply:
@norunners as long as the rest of the lifecycle hooks get automatically triggered without having for vstore to keep up, then we're good 👍